stmhal/usbd_hid_interface.c

```diff
@@ -0,0 +1,109 @@
+  *
+  * Unless required by applicable law or agreed to in writing, software 
+  * distributed under the License is distributed on an "AS IS" BASIS, 
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  *
+  ******************************************************************************
+  */
+
+/* Includes ------------------------------------------------------------------*/
+
+#include <stdint.h>
+
+#include "usbd_cdc_msc_hid.h"
+#include "usbd_hid_interface.h"
+
+#include "py/obj.h"
+#include "irq.h"
+#include "usb.h"
+
+/* Private variables ---------------------------------------------------------*/
+
+static __IO uint8_t dev_is_connected = 0; // indicates if we are connected
+
+#define HID_RX_PACKET_SIZE 64
+
+static uint8_t buffer[2][HID_RX_PACKET_SIZE]; // pair of buffers to read individual packets into
+static int8_t current_read_buffer = 0; // which buffer to read from
+static uint32_t last_read_len; // length of last read
+static int8_t current_write_buffer = 0; // which buffer to write to
+
+
+/* Private function prototypes -----------------------------------------------*/
+static int8_t HID_Itf_Receive  (uint8_t* pbuf, uint32_t Len);
+
+const USBD_HID_ItfTypeDef USBD_HID_fops = {
+    HID_Itf_Receive
+};
+
+/**
+  * @brief  HID_Itf_Receive
+  *         Data received over USB OUT endpoint is processed here.
+  * @param  Buf: Buffer of data received
+  * @param  Len: Number of data received (in bytes)
+  * @retval Result of the opeartion: USBD_OK if all operations are OK else USBD_FAIL
+  * @note   The buffer we are passed here is just UserRxBuffer, so we are
+  *         free to modify it.
+  */
+static int8_t HID_Itf_Receive(uint8_t* Buf, uint32_t Len) {
+    current_write_buffer = !current_write_buffer;
+    last_read_len = Len;
+    // initiate next USB packet transfer, to append to existing data in buffer
+    USBD_HID_SetRxBuffer(&hUSBDDevice, buffer[current_write_buffer]);
+    USBD_HID_ReceivePacket(&hUSBDDevice);
+
+    return USBD_OK;
+}
+
+// timout in milliseconds.
+// Returns number of bytes read from the device.
+int USBD_HID_Rx(uint8_t *buf, uint32_t len, uint32_t timeout) {
+    // Wait until we have at least 1 byte to read
+    uint32_t start = HAL_GetTick();
+    while (current_read_buffer == current_write_buffer) {
+        // Wraparound of tick is taken care of by 2's complement arithmetic.
+        if (HAL_GetTick() - start >= timeout) {
+            // timeout
+            return 0;
+        }
+        if (query_irq() == IRQ_STATE_DISABLED) {
+            // IRQs disabled so buffer will never be filled; return immediately
+            return 0;
+        }
+        __WFI(); // enter sleep mode, waiting for interrupt
+    }
+
+    // Copy bytes from device to user buffer
+    memcpy(buf, buffer[current_read_buffer], last_read_len);
```

Comment:
The len argument to this function is not used at all, in particular this line could overrun buf if buf is too short.